beacon-chain/slasher/chunks.go

```diff
@@ -0,0 +1,553 @@
+// chunk_idx = (epoch % H) / C = (4 % 2) / 2 = 0
+//
+//
+//                                     val0    val1    val2
+//                                     {  }    {  }    {  }
+//   chunk_0_for_validators_0_to_3 = [[2, 2], [2, 2], [2, 2]]
+//                                     |
+//                                     |-> epoch 4 for validator 0
+//
+// Next up, we proceed with the update process for validator index 0, starting epoch 4
+// all the way down to epoch 2. We will need to go down the array as far as we can get. If the
+// lowest epoch we need to update is < the lowest epoch of a chunk, we need to proceed to
+// a different chunk index.
+//
+// Once we finish updating a chunk, we need to move on to the next chunk. This function
+// returns a boolean named keepGoing which allows the caller to determine if we should
+// continue and update another chunk index. We stop whenever we reach the min epoch we need
+// to update, in our example, we stop at 2, which is still part of chunk 0, so no need
+// to jump to another min span chunks slice to perform updates.
+func (m *MinSpanChunksSlice) Update(
+	args *chunkUpdateArgs,
+	validatorIndex types.ValidatorIndex,
+	startEpoch,
+	newTargetEpoch types.Epoch,
+) (keepGoing bool, err error) {
+	// The lowest epoch we need to update. This is a sliding window from (current epoch - H) where
+	// H is the history length a min span for a validator stores.
+	var minEpoch types.Epoch
+	if args.currentEpoch > (m.params.historyLength - 1) {
+		minEpoch = args.currentEpoch - (m.params.historyLength - 1)
+	}
+	epochInChunk := startEpoch
+	// We go down the chunk for the validator, updating every value starting at start_epoch down to min_epoch.
+	// As long as the epoch, e, in the same chunk index and e >= min_epoch, we proceed with
+	// a for loop.
+	for m.params.chunkIndex(epochInChunk) == args.chunkIndex && epochInChunk >= minEpoch {
+		var chunkTarget types.Epoch
+		chunkTarget, err = chunkDataAtEpoch(m.params, m.data, validatorIndex, epochInChunk)
+		if err != nil {
+			err = errors.Wrapf(err, "could not get chunk data at epoch %d", epochInChunk)
+			return
+		}
+		// If the newly incoming value is < the existing value, we update
+		// the data in the min span to meet with its definition.
+		if newTargetEpoch < chunkTarget {
+			if err = setChunkDataAtEpoch(m.params, m.data, validatorIndex, epochInChunk, newTargetEpoch); err != nil {
+				err = errors.Wrapf(err, "could not set chunk data at epoch %d", epochInChunk)
+				return
+			}
+		} else {
+			// We can stop because spans are guaranteed to be minimums and
+			// if we did not meet the minimum condition, there is nothing to update.
+			return
+		}
+		// We decrease our epoch index variable as long as it is > 0.
+		if epochInChunk > 0 {
+			epochInChunk -= 1
+		}
+	}
+	// We should keep going and update the previous chunk if we are yet to reach
+	// the minimum epoch required for the update procedure.
+	keepGoing = epochInChunk >= minEpoch
```

Comment:
Something to consider: create a `type KeepGoing bool` instead of using a named parameter.